src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -31,6 +31,72 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region InodeTracker
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<UInt64, Dictionary<UInt64, bool>> _visitations;
+
+        /// <summary>
+        /// Construct a new InodeTracker
+        /// </summary>
+        public InodeTracker()
+        {
+            _visitations = new Dictionary<UInt64, Dictionary<UInt64, bool>>();
+        }
+
+        /// <summary>
+        /// Determine if a path has been visited.
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be checked.</param>
+        /// <returns>True if the path has been visited, false otherwise</returns>
+        public bool Visited(string path)
+        {
+            UInt64 dev;
+            UInt64 inode;
+
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out dev, out inode))
+            {
+                if (_visitations.ContainsKey(dev))
+                {
+                    return _visitations[dev].ContainsKey(inode);
```

Comment:
>C#7 pattern

We have no strict documentation. Different maintainers may have different requirements. My understanding: New patterns are welcome for the new code. Existing patterns (not relevant to the PR) should not be changed without maintainer's requiring.

> the nested dictionary approach may be faster.

There, the main answer has recently been updated and there is a reference to [gist with the tests](https://gist.github.com/ljw1004/61bc96700d0b03c17cf83dbb51437a69). The conclusion is now Tuple keys is fast.
We both think that performance is important here. So maybe it makes sense to spend time testing cases.
Another thought is, in most cases, the user will have one volume. We could use one dictionary until we find a reference to another volume. Although it will complicate the code and it's not clear how much we win. Ooops, for other volume we can make separate recursive `Dir()` call - in the case we have to switch Dictionary (tracker) only on "top" level when we meet new volume. In other words, we will have two dictionaries but they will not be nested and the number of comparisons will be less - we'll be faster.

And maybe rename `Dir()` in something like `DirectoryVisitor()` ?